src/nifs/protogalaxy/mod.rs

```diff
@@ -139,3 +149,135 @@ impl<C: CurveAffine> MultifoldingScheme<C> for ProtoGalaxy<C> {
         todo!()
     }
 }
+
+/// given a vector of {w_i}, here w_i is a vector
+/// compute weighted sum: sum_i L_i(gamma)c_i
+/// Here w_i is vector of challenges
+/// # Parameters
+///
+/// - `gamma` - eval Lagrange polynomials at this point
+/// - `log_n` - `log2(n+1)`, where `n` - number of instances to be folded
+/// - `ws` - `{w_0,..., w_n}`, w_i is vector of F
+///
+/// # Result
+///
+/// the vector of weighted sum
+/// `L_0(gamma)w_0 + L_1(gamma)w_1, ..., L_n(gamma)w_n`
+fn fold_challenges<F: PrimeField>(gamma: F, log_n: usize, ws: &[&[F]]) -> Vec<F> {
+    let ll: Vec<F> = iter_eval_lagrange_polynomials_for_cyclic_group(gamma, log_n).collect();
+    (0..ws[0].len())
+        .into_par_iter()
+        .map(|idx| {
+            ll.iter()
+                .zip(ws.iter())
+                .map(|(l_i, w_i)| *l_i * w_i[idx])
+                .fold(F::ZERO, |sum, val| sum + val)
+        })
+        .collect()
+}
+
+/// given a vector of {w_i}, here witness w_i is vector of vector
+/// compute weighted sum: sum_i L_i(gamma)w_i
+/// Here w_i is witness, vector of vector
+/// # Parameters
+///
+/// - `gamma` - eval Lagrange polynomials at this point
+/// - `log_n` - `log2(n+1)`, where `n` - number of instances to be folded
+/// - `ws` - `{w_0,..., w_n}`, w_i is vector of vector of F
+///
+/// # Result
+///
+/// the vector of weighted sum
+/// `L_0(gamma)w_0 + L_1(gamma)w_1, ..., L_n(gamma)w_n`
+fn fold_witnesses<F: PrimeField>(gamma: F, log_n: usize, wss: &[&[Vec<F>]]) -> Vec<Vec<F>> {
+    let ll: Vec<F> = iter_eval_lagrange_polynomials_for_cyclic_group(gamma, log_n).collect();
+    (0..wss[0].len())
+        .into_par_iter()
+        .map(|ii| {
+            (0..wss[0][ii].len())
+                .into_par_iter()
```

Comment:
It is undesirable to use rayon like this.

Each column will parallely spawn more rayon-task and it is not efficient

Considering the number of columns, the simple solution here is to iterate parallel only on rows.

The next step could be to count all `PlonkWitness` for all challenges in one paralel pass